src/graphene.cpp

```diff
@@ -0,0 +1,1562 @@
+        pfrom->grapheneBlockHashes[pfrom->grapheneMapHashOrderIndex[cheapHash]] = hash;
+    }
+
+    LOG(GRAPHENE, "Got %d Re-requested txs from peer=%s\n", grapheneBlockTx.vMissingTx.size(), pfrom->GetLogName());
+
+    // At this point we should have all the full hashes in the block. Check that the merkle
+    // root in the block header matches the merkel root calculated from the hashes provided.
+    bool mutated;
+    uint256 merkleroot = ComputeMerkleRoot(pfrom->grapheneBlockHashes, &mutated);
+    if (pfrom->grapheneBlock.hashMerkleRoot != merkleroot || mutated)
+    {
+        graphenedata.ClearGrapheneBlockData(pfrom, inv.hash);
+
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Merkle root for %s does not match computed merkle root, peer=%s", inv.hash.ToString(),
+            pfrom->GetLogName());
+    }
+    LOG(GRAPHENE, "Merkle Root check passed for %s peer=%s\n", inv.hash.ToString(), pfrom->GetLogName());
+
+    // Xpress Validation - only perform xval if the chaintip matches the last blockhash in the graphene block
+    bool fXVal;
+    {
+        LOCK(cs_main);
+        fXVal = (pfrom->grapheneBlock.hashPrevBlock == chainActive.Tip()->GetBlockHash()) ? true : false;
+    }
+
+    int missingCount = 0;
+    int unnecessaryCount = 0;
+    // Look for each transaction in our various pools and buffers.
+    // With grapheneBlocks the vTxHashes contains only the first 8 bytes of the tx hash.
+    {
+        LOCK2(orphanpool.cs, cs_xval);
+        if (!ReconstructBlock(pfrom, fXVal, missingCount, unnecessaryCount))
+            return false;
+    }
+
+    // If we're still missing transactions then bail out and just request the full block. This should never
+    // happen unless we're under some kind of attack or somehow we lost transactions out of our memory pool
+    // while we were retreiving missing transactions.
+    if (missingCount > 0)
+    {
+        // Since we can't process this graphene block then clear out the data from memory
+        graphenedata.ClearGrapheneBlockData(pfrom, inv.hash);
+
+        std::vector<CInv> vGetData;
+        vGetData.push_back(CInv(MSG_BLOCK, grapheneBlockTx.blockhash));
+        pfrom->PushMessage(NetMsgType::GETDATA, vGetData);
+        return error("Still missing transactions after reconstructing block, peer=%s: re-requesting a full block",
+            pfrom->GetLogName());
+    }
+    else
+    {
+        // We have all the transactions now that are in this block: try to reassemble and process.
+        CInv inv(CInv(MSG_BLOCK, grapheneBlockTx.blockhash));
+
+        // for compression statistics, we have to add up the size of grapheneblock and the re-requested grapheneBlockTx.
+        int nSizeGrapheneBlockTx = msgSize;
+        int blockSize = ::GetSerializeSize(pfrom->grapheneBlock, SER_NETWORK, CBlock::CURRENT_VERSION);
+        LOG(GRAPHENE, "Reassembled xblocktx for %s (%d bytes). Message was %d bytes (graphene block) and %d bytes "
```

Comment:
-LOG(GRAPHENE, "Reassembled xblocktx ....
+LOG(GRAPHENE, "Reassembled grapheneBlockTx ?

"xblocktx" needs to be whatever the graphene equivalent is here..

Reply:
Done.